## 數據程式設計 作業一
---
**<font color=tomato>希望大家真的學會寫程式，所以請自行寫作，可以跟同學討論但不要明顯地抄襲，否則兩者通通不給分！</font> \
明顯地抄襲: 包含僅改變變數的名稱 或局部的內容！**

請在題目的下面新增一個cell來撰寫作業的程式內容，最後再將此作答完成的ipynb檔，\
存成 "<font color=blue>你的學號-姓名-homework-1.ipynb</font>" 上傳繳交至此課程的北科i學園。
---
**本次作業共有兩題，另有一題加分題(<font color=tomato>可做可不做</font>)！**


### 題目一

請寫一個 Python 程式，能將任意給定的學生的資料([資料請由此下載(右鍵另存)](http://140.129.118.16/~richwang/exercise/data/Part.txt))， \
分別使用 
1. HTML 表格，以及 
2. 下列示範的文字的形式印出來。

<pre><font face="細明體" color='blue'>
+--------+----------+----------+------------------+
|  姓名  |   學號   |   系別   | email            | 
+--------+----------+----------+------------------+      
| 葉聿欣 | 11112345 |  資工二  | mary@gmail.com   |
+--------+----------+----------+------------------+
| 李清蓁 | 11154246 | 電機三甲 | fish@ntut.edu.tw |
+--------+----------+----------+------------------+
</font></pre>
```
輸出內容，除 email 欄位靠左對齊外，其餘欄位則請設定成置中顯示，欄位跟左右邊界至少有一格的留白。
```

In [10]:
fname = "./Part.txt"
entries = []

with open(fname, "r", encoding='utf-8') as fp:
    lines = fp.read()
    lines = lines.split('\n')

for line in lines:
    entry = line.split(',')
    entries.append(entry)

chart = ""
html_chart = "<table>\n"

for entry in entries:
    html_chart += "\t<tr>\n"
    for item in entry:
        html_chart += f"\t\t<th>{item}</th>\n"
    html_chart += "\t</tr>\n"

html_chart += "</table>"

print(html_chart)
with open("part.html", "w", encoding='utf-8') as fp:
    fp.write(html_chart)

<table>
	<tr>
		<th>葉聿欣</th>
		<th>11112345</th>
		<th>資工二</th>
		<th>mary@gmail.com</th>
	</tr>
	<tr>
		<th>李清蓁</th>
		<th>11154246</th>
		<th>智動四</th>
		<th>fish@ntut.edu.tw </th>
	</tr>
	<tr>
		<th>趙欣勇</th>
		<th>11133250</th>
		<th>電機三甲</th>
		<th>brave@ntut.edu.tw </th>
	</tr>
	<tr>
		<th>劉清</th>
		<th>11161022</th>
		<th>資財四甲</th>
		<th>clear@ntut.edu.tw </th>
	</tr>
	<tr>
		<th>歐陽素素</th>
		<th>11155135</th>
		<th>電機四丙</th>
		<th>swing@ntut.edu.tw </th>
	</tr>
</table>


html chart:
<table>
	<tr>
		<th>葉聿欣</th>
		<th>11112345</th>
		<th>資工二</th>
		<th>mary@gmail.com</th>
	</tr>
	<tr>
		<th>李清蓁</th>
		<th>11154246</th>
		<th>智動四</th>
		<th>fish@ntut.edu.tw </th>
	</tr>
	<tr>
		<th>趙欣勇</th>
		<th>11133250</th>
		<th>電機三甲</th>
		<th>brave@ntut.edu.tw </th>
	</tr>
	<tr>
		<th>劉清</th>
		<th>11161022</th>
		<th>資財四甲</th>
		<th>clear@ntut.edu.tw </th>
	</tr>
	<tr>
		<th>歐陽素素</th>
		<th>11155135</th>
		<th>電機四丙</th>
		<th>swing@ntut.edu.tw </th>
	</tr>
</table>


### 題目二

**請寫一個Python程式，可以計算任意矩陣(mxn)的基本運算: 加法，減法與乘法 (+,-,*)。** \
如果不能計算時，請輸出訊息告知使用者！ \
$A_{m \times n} = [a_{ij}]_{m \times n}$: 表示 m 列(rows) n 行(columns)的矩陣。 \
$A_{m \times n} = [a_{ij}]_{m \times n}=
\begin{vmatrix}
a_{11} & a_{12} & \dots & a_{1n} \\
a_{21} & a_{22} & \dots & a_{2n} \\
\vdots & \cdots & \ddots & \vdots \\
a_{m1} & a_{m2} & \dots & a_{mn}
\end{vmatrix}$

$若已知 A=[a_{ij}]_{m \times n}, B=[b_{ij}]_{n \times p},$
$則 A_{m \times n}*B_{n \times p}=C_{m \times p}=[c_{ij}]_{m \times p}$ \
$$c_{ij} = \sum_{k=1}^{n}{a_{ik}b_{kj}}, \; for \: 1 <= i <= m, \: 1 <= j <= p$$

```python
class Matrix:
    def __init__(self, row, col, values):
        # values: 型態是一維 list，裡面存的是矩陣元素的內容，順序為先列後行。 
        self.row = row
        self.col = col
        # 矩陣元素請以二維串列來儲存，請先完成下列二維串列(mat)的初值設定。
        self.mat = [[]]
        
    # 矩陣運算前，需要檢視行列是否符合計算規定！
    def __add__(self, other):
        pass
    def __sub__(self, other):
        pass
    
    # 乘法包含矩陣相乘，與矩陣與數值相乘。
    # 矩陣*矩陣, 矩陣*數值, 數值*矩陣，三種都要能進行運算！
    def __mul__(self, other):
        pass
    def __rmul__(self, other):
        pass
    
    # 請為矩陣兩側加上'|'，原則上將元素當成三位數來列印。
    def __show__(self):
        pass
```
若已知矩陣
$A=\begin{vmatrix}
1 & 2 & 3 & 4\\
5 & 6 & 7 & 8\\
9 & 8 & 7 & 6
\end{vmatrix}_{3 \times 4}, 
B=\begin{vmatrix}
3 & 4 & 6\\
2 & 5 & 7\\
6 & 8 & 9\\
8 & 4 & 3
\end{vmatrix}_{4 \times 3},
C=\begin{vmatrix}
1 & 3 & 5 & 7\\
2 & 4 & 6 & 8\\
2 & 0 & 2 & 3
\end{vmatrix}_{3 \times 4}$, \
請計算下列矩陣運算式的結果:
1. $A+B$
2. $A+C$
3. $A*B$
4. $B*A$
5. $A*C$
6. $B*C$
8. $A*3$
9. $5*B$

In [3]:
class Matrix:
    def __init__(self, row:int, col:int, values:list):
        # values: 型態是一維 list，裡面存的是矩陣元素的內容，順序為先列後行。 
        self.row = row
        self.col = col
        # 矩陣元素請以二維串列來儲存，請先完成下列二維串列(mat)的初值設定。
        self.mat = []
        i = 0
        new_row = []
        for element in values:
            new_row.append(element)
            i += 1
            if(i%col == 0):
                self.mat.append(new_row)
                new_row = []

        
    # 矩陣運算前，需要檢視行列是否符合計算規定！
    def __add__(self, other):
        if(other.row != self.row or self.col != other.col):
            print("It doesn't match the matrix addition rules.")
        else:
            new_mat = []
            for i in range(self.row):
                for j in range(self.col):
                    new_mat.append(self.mat[i][j] + other.mat[i][j])
            return Matrix(self.row, self.col, new_mat)

    def __sub__(self, other):
        if(other.row != self.row or other.col != self.row):
            print("It doesn't match the matrix subtraction rules.")
        else:
            new_mat = []
            for i in range(self.row):
                for j in range(self.col):
                    new_mat.append(self.mat[i][j] - other.mat[i][j])
            return Matrix(self.row, self.col, new_mat)

        return Matrix(self.row, self.col, new_mat)
    
    # 乘法包含矩陣相乘，與矩陣與數值相乘。
    # 矩陣*矩陣, 矩陣*數值, 數值*矩陣，三種都要能進行運算！
    def __mul__(self, other):
        if(self.col != other.row):
            return("Cannot Multiply.")
        else:
            new_mat = []
            element = 0
            for i in range(self.row):
                for j in range(self.col):
                    element += self.mat[i][j]*other.mat[j][i]
                new_mat.append(element)
            return(new_mat)

    def __mul__(self, coef:int):
        pass

    def __rmul__(self, other):
        pass
    
    # 請為矩陣兩側加上'|'，原則上將元素當成三位數來列印。
    def __show__(self):
        pass

    def show(self):
        print(self.mat)

A = Matrix(3, 4, [1, 2, 3, 4, 5, 6, 7, 8, 9, 8, 7, 6])
B = Matrix(4, 3, [3, 4, 6, 2, 5, 7, 6, 8, 9, 8, 4, 3])
C = Matrix(3, 4, [1, 3, 5, 7, 2, 4, 6, 8, 2, 0, 2, 3])
print(A+B)
(A+C).show()



It doesn't match the matrix addition rules.
None
[[2, 5, 8, 11], [7, 10, 13, 16], [11, 8, 9, 9]]


**<font color=red>加分題！</font>請寫 Python 程式，並配合 os.listdir()，以下列格式列出任意起始目錄(ex., my_data)其下的目錄樹！** \
不同層次的最終目錄或檔案，請以 @ 來表示，其餘符號請觀察下列的示範內容，自行理解給出。
<pre>
my_data
  |--data
  |  |--dat
  |  |  @--video
  |  @--txt
  |--gif
  |--jpg
  |--png
  @--python
     |--ipynb
     @--py
</pre>